### ISSUE TITLE
Caching for improvements in kernel startup

### ISSUE THREAD

**Issue by DonJayamanne** [microsoft/vscode-jupyter#6702]
Today Jupyter extension calls API in VS Code to determine:
* If IPyKernel is installed
* Determine if the version of IPYKernel satisfies some requirements (>=6.0)
	* Used in debugging
* etc

**Suggestion: Cache in Python extension**
* When getting the version, we use the code `import xyz; print(xyz.__version__)`
	* I'm suggesting we also get the `__file__`
* From the __file__ we can determine the directory where this module is installed, get the last modified date/time of that & cache it
* When we calling this code in the same  VS Code session, we can check if that folder has been modified, if not, then return the cached version (no we avoid running a whole process all over, just a simple disc IO)

Optionally, we can ignore the cache when restarting VS Code (i.e. the cache is only per-session).
This way, if something doesn't work properly, all we need to do is ask the user to restart VS Code (this is a worst case scenario).

**Suggestion: Cache in Python extension**
* Optionally we could perform this same algorithm in the Jupyter extension. Instead of checking the last modified date of the individual package folder, we could check the last modified date of the Python environment folder. 
* If user installs/updates/uninstalls packages in a python environment, then the python environment folders modified date changes, hence we can invalidate the cache easily at our end.
* Ie.. cache key = Interpreter Path + Last Modifeid date of interpreter folder

Caching in Jupyter extension is super easy & safe, with great benefits for us.

@rchiodo @DavidKutu @joyceerhl /cc

**Comment by DonJayamanne** [comment]
**Solution Implemented:**
* If we have detected ipykernel for an interprerter once before, avoid looking for this before we start kernels.
* If user subsequently uninstalls ipykernel, then we'll detect this when kernel fails to start.

**Comment by DonJayamanne** [comment]
Duplicate of https://github.com/microsoft/vscode-jupyter/issues/8196

**Comment by DonJayamanne** [comment]
**Testing**

* Open VS Code stable, create a notebook, run a cell with `print` statement
* Keep the notebook opened.
* Reload VS Code and run the cell immediately.
* Check the logs in `Jupyter` output panel to determine the time from activation to time take to run the cell. 

Try the same in VS Code insiders, (remember to reload)
* The times should be faster in VS Code.

**Note:** 
* Things are faster only when running a cell for an existing notebook that you ran previously.
* You might want to try a conda environment (on windows) as they are slower (Mac/Linux is faster & the time difference isn't as noticeable)
